atomic_reactor/utils/resource_manager.py

```diff
@@ -0,0 +1,185 @@
+                    logger.debug("%s slot %s is %s", meta['hostname'], i, state)
+                    self.slots.append({'sid': i, 'state': state, 'prid': prid})
+                    i += 1
+                return True
+        except Exception as err:
+            logger.debug(err)
+            logger.error("Availiability check failed fo host: %s", self.hostname)
+            return False
+
+    def _write_free_slot(self, slot: int):
+        cmd = f'echo \'0\' > {ROOT_PATH}slot_{slot}.json'
+        data = self._ssh_run_remote_command(cmd)
+        return data
+
+    def unlock(self, prid: str):
+        logger.debug('Unlocking')
+        if self.check_avail():
+            for slot in self.slots:
+                if slot.get('prid', '') == prid:
+                    retries = 0
+                    while retries < 3:
+                        ret = self._write_free_slot(slot.get('sid'))
+                        if ret == b'':
+                            logger.info('Unlocked')
+                            return True
+                        else:
+                            logger.debug('Unlocking failed, retrying')
+                            retries += 1
+        return False
+
+    def lock(self, slot: int, prid: str):
+        logger.debug("Aquiring lock")
+        payload = json.dumps({"prid": prid, "locked": str(datetime.now())})
+        retries = 0
+        while retries < 3:
+            filename = os.path.join(ROOT_PATH, f'slot_{slot}.json')
+            _cmd = 'if grep -qw 0 {0} ; then echo {1} > {0} ;else exit 1; fi'.format(
+                    filename,
+                    quote(payload)
+                    )
+            cmd = 'flock {} -c {}; echo $?'.format(filename, quote(_cmd))
+            logger.debug(cmd)
+            ret = self._ssh_run_remote_command(cmd)
+            if ret == b'0\n':
+                logger.info('Locked')
+                return True
+            else:
+                logger.debug('Locking failed, retrying')
+                retries += 1
+        logger.info('Locking failed')
+        return False
+
+    def _ssh_run_remote_command(self, cmd: str):
+        ssh_client = paramiko.SSHClient()
+        ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        ssh_client.connect(hostname=self.hostname,
+                           username=self.username)
+        cmd_out = ssh_client.exec_command(cmd)
+        out = cmd_out[1].read()
+        error = cmd_out[2].read()
+        if error:
+            raise Exception('There was an error pulling the runtime: {}'.format(error))
+        ssh_client.close()
+        return out
+
+
+class ResourceManager:
+    '''
+    Manages build resources
+    '''
+
+    def __init__(self, hosts: list[Host]):
+        self.hosts = hosts
+
+    def find(self) -> list[Host]:
+        try:
+            resources = []
+            for host in self.hosts:
+                host.check_avail()
+                if len(host.free_slots) > 0:
```

Comment:
You need to call the `check_avail()` method, otherwise, free_slots will be not up to date. Isn't it better to refactor `check_avail` method to return relevant data?  Keeping free_slots seems to be useless.